### ISSUE TITLE
Add a method splitting a JoinKeys object into multiple JoinKeys objects: each corresponding to a single dataname

### ISSUE THREAD

**Issue by denisovan31415** [insightsengineering/teal#125]
see https://github.com/insightsengineering/teal/pull/121#pullrequestreview-729334532

--------------------------
> It doesn't appear to me to be the right direction, with hindsight of course. Since a Dataset or DatasetConnnector object should be decoupled and encapsulated, it really shouldn't have to deal with join_keys

Wholeheartedly agree, but the aforementioned obstacle prevents it. The missing piece is the initialization of an "empty" Dataset object in the DatasetConnector constructor to exploit its mutation and join keys logic in Connector, but it's out of the scope of this issue.

I would also say that join keys in DatasetConnector is not the worst thing imaginable - the logic of mutating and holding keys is encapsulated inside JoinKeys. We see the complexity increase in Dataset and DatasetConnector is minimal.

The question is: what to do with the join keys in RelationalData and I have another proposition to you, @junlueZH .

Due to how our API is constructed we end up with a JoinKeys object inside of RelationalData, so let's make the best out of this happenstance. Let's keep the logic dealing with join keys inside JoinKeys class. At first glance I see that we need the following functionality:
* split a JoinKeys object into multiple JoinKeys objects - each corresponding to a single dataset present in the original JoinKeys
I propose it as a method of JoinKeys with a signature JoinKeys -> list(...) where ... are JoinKeys objects. The idea is RelationalData calls this method, then distributes the appropriate JoinKeys to Dataset and DatasetConnector objects.

We also need to still support mutating the join keys, an issue tied to the question of whether to store the original, "joined" join keys in RelationalData or not. 
Advantages of storing a "joined" JoinKeys object in RelationalData: mutation is laughingly easy to support - just mutate the joined JoinKeys, call the new method and redistribute the JoinKeys again
Disadvantages: the encapsulation is broken, it's harder to maintain, harder to understand and also it duplicates the information for no good reason except ease of implementing one method

So I would go for storing the JoinKeys object only in Dataset and DatasetConnector object. But then we have no "joined" JoinKeys and we need to support get_join_keys on RelationalData, so this also creates a need for the next functionality:
* we need to be able to merge multiple JoinKeys object into one
The signature of this function would be this list(...) -> JoinKeys where ... are JoinKeys objects. Make sure to use as much of the public API of JoinKeys as possible in there. E.g. adding keys to JoinKeys deals with duplicates on its own:
```r
x <- join_keys(
  join_key("dataset_A", "dataset_B", c("col_1" = "col_a")),
  join_key("dataset_B", "dataset_A", c("col_a" = "col_1")),
  join_key("dataset_A", "dataset_C", c("col_2" = "col_x", "col_3" = "col_y"))
)
```
something like this passes and has only JoinKey for datasets A and B.

This function would be called by RelationalData$get_join_keys and it would solve our problems.

To sum it up:
* delegate the logic of dealing with JoinKeys to the JoinKeys class itself
* don't forget to add generics for Dataset and DatasetConnector to the mutate_join_keys S3 method
